proxy/src/coco/peer.rs

```diff
@@ -32,316 +39,400 @@ pub struct UserRevisions {
-
-        let storage = peer.storage().reopen()?;
-
-        if storage.has_urn(&urn)? {
-            return Err(error::Error::EntityExists(urn));
-        } else {
-            let repo = storage.create_repo(&meta)?;
-            repo.set_rad_self(librad::git::storage::RadSelfSpec::Urn(owner.urn()))?;
+    /// Get the project found at `project_urn`.
+    ///
+    /// # Errors
+    ///
+    ///   * Parsing the `project_urn` fails.
+    ///   * Resolving the project fails.
+    pub fn get_project(
+        &self,
+        urn: &RadUrn,
+    ) -> Result<project::Project<entity::Draft>, error::Error> {
+        let api = self.peer_api.lock().expect("unable to acquire lock");
+        let storage = api.storage().reopen()?;
+
+        Ok(storage.metadata(urn)?)
+    }
+
+    /// Get the user found at `urn`.
+    ///
+    /// # Errors
+    ///
+    ///   * Resolving the project fails.
+    ///   * Could not successfully acquire a lock to the API.
+    pub fn get_user(&self, urn: &RadUrn) -> Result<user::User<entity::Draft>, error::Error> {
+        let api = self.peer_api.lock().expect("unable to acquire lock");
+        let storage = api.storage().reopen()?;
+
+        Ok(storage.metadata(urn)?)
+    }
+
+    /// Get a repo browser for a project.
+    ///
+    /// # Errors
+    ///
+    /// The function will result in an error if the mutex guard was poisoned. See
+    /// [`std::sync::Mutex::lock`] for further details.
+    pub fn with_browser<F, T>(&self, project_urn: &RadUrn, callback: F) -> Result<T, error::Error>
+    where
+        F: Send + FnOnce(&mut git::Browser) -> Result<T, error::Error>,
+    {
+        let git_dir = {
+            let paths = self.paths();
+            paths.git_dir().join("")
+        };
+
+        let project = self.get_project(project_urn)?;
+        let default_branch = git::Branch::local(project.default_branch());
+        let repo = git::Repository::new(git_dir)?;
+        let namespace = git::Namespace::try_from(project.urn().id.to_string().as_str())?;
+        let mut browser = git::Browser::new_with_namespace(&repo, &namespace, default_branch)?;
+
+        callback(&mut browser)
+    }
+
+    /// Initialize a [`project::Project`] that is owned by the `owner`.
+    /// This kicks off the history of the project, tracked by `librad`'s mono-repo.
+    ///
+    /// # Errors
+    ///
+    /// Will error if:
+    ///     * The signing of the project metadata fails.
+    ///     * The interaction with `librad` [`librad::git::storage::Storage`] fails.
+    #[allow(clippy::needless_pass_by_value)] // We don't want to keep `SecretKey` in memory.
+    pub fn init_project(
+        &self,
+        key: keys::SecretKey,
+        owner: &User,
+        path: impl AsRef<std::path::Path> + Send,
+        name: &str,
+        description: &str,
+        default_branch: &str,
+    ) -> Result<project::Project<entity::Draft>, error::Error> {
+        let api = self.peer_api.lock().expect("unable to acquire lock");
```

Comment:
maybe this could be extracted as well, also to make sure that the error message is consistent and to avoid repetition.

Reply:
Definitely blood. The issue with extraction anything sync related is that these lock mechanisms hinge on drop semantics. And it is already hard enough to debug all these cases where we deadlock at the moment. For now it's safer to handle locking explicitly in the method bodies. We could maybe introduce an error for the consistent error messaging.